tests/test_spectrograph.py

```diff
@@ -298,6 +355,146 @@ def test_get_status_getAnalogIn_fails(self):
+        self.patch.return_value.AVS_GetScopeData.return_value = AvsReturnCode.ERR_INVALID_MEAS_DATA.value
+
+        spec = AvsFiberSpectrograph()
+        with self.assertRaisesRegex(AvsReturnError, "GetScopeData"):
+            await spec.expose(duration)
+        self.patch.return_value.AVS_PrepareMeasure.assert_called_once()
+        self.patch.return_value.AVS_Measure.assert_called_once_with(self.handle, 0, 1)
+        self.assertEqual(self.patch.return_value.AVS_PollScan.call_count, 4)
+
+    async def test_stop_exposure(self):
+        """Test that `stop_exposure` makes ends the active `expose`."""
+        duration = 5  # seconds
+        spec = AvsFiberSpectrograph()
+
+        t0 = time.perf_counter()
+        task = asyncio.create_task(spec.expose(duration))
+        await asyncio.sleep(0.1)  # give the event loop time to start
+        spec.stop_exposure()
+        await task
+        t1 = time.perf_counter()
+
+        self.assertLess(t1 - t0, 1)  # cancelling the task should make it end much sooner than the duration
+        self.patch.return_value.AVS_StopMeasure.assert_called_with(self.handle)
+        self.assertIsNone(task.result())
+
+    async def test_stop_exposure_no_expose_running(self):
+        """Test that stop_exposure does nothing if there is no active
+        `expose` command.
+        """
+        spec = AvsFiberSpectrograph()
+        spec.stop_exposure()
+        self.patch.return_value.AVS_StopMeasure.assert_not_called()
+
+    async def test_stop_exposure_fails(self):
+        """Test `AVS_StopMeasure` returning an error: the existing exposure
+        task should be cancelled, but `stop_exposure` should also raise."""
+        duration = 5  # seconds
+        self.patch.return_value.AVS_StopMeasure.return_value = AvsReturnCode.ERR_TIMEOUT.value
+        spec = AvsFiberSpectrograph()
+
+        t0 = time.perf_counter()
+        task = asyncio.create_task(spec.expose(duration))
+        await asyncio.sleep(0.1)  # give the event loop time to start
+        with self.assertRaisesRegex(AvsReturnError, "StopMeasure"):
+            spec.stop_exposure()
+        await task
+        t1 = time.perf_counter()
+
+        self.assertLess(t1 - t0, 1)  # cancelling the task should make it end much sooner than the duration
+        self.patch.return_value.AVS_StopMeasure.assert_called_with(self.handle)
+        self.assertIsNone(task.result())
+
+    async def test_disconnect_active_exposure(self):
+        """Test that disconnecting cancels an active exposure."""
+        duration = 5  # seconds
+        spec = AvsFiberSpectrograph()
+
+        t0 = time.perf_counter()
+        task = asyncio.create_task(spec.expose(duration))
+        await asyncio.sleep(0.1)  # give the event loop time to start
+        spec.disconnect()
+        await task
+        t1 = time.perf_counter()
+
+        self.assertLess(t1 - t0, 1)  # cancelling the task should make it end much sooner than the duration
```

Comment:
I'm feeling a bit uncomfortable with all the hard coded numbers. 1 here is >> 0.1 you use to let the event loop and << duration = 5. Is there a clean way to express this `duration` instead of `1`

0.1 is probably fine hard-coded (though a global is also nice). Even `await asyncio.sleep(0)` will free the event loop, but like the fact that you wait a bit before the disconnect/stop/whatever.

Reply:
I've added some more tests that use other durations too, for various reasons. I don't know how long to consider a "successful stop" of a 5s exposure: it will be a bit longer than the `sleep`, but I'd rather make it long enough to ensure that any extra time spent waiting due to e.g. a busy CPU doesn't cause a failing test.